In [1]:
import pandas as pd


In [2]:
df = pd.read_csv('games.csv')
df.shape

(26651, 21)

In [3]:
df.head()

,GAME_DATE_EST,GAME_ID,GAME_STATUS_TEXT,HOME_TEAM_ID,VISITOR_TEAM_ID,SEASON,TEAM_ID_home,PTS_home,FG_PCT_home,FT_PCT_home,...,AST_home,REB_home,TEAM_ID_away,PTS_away,FG_PCT_away,FT_PCT_away,FG3_PCT_away,AST_away,REB_away,HOME_TEAM_WINS
0,2022-12-22,22200477,Final,1610612740,1610612759,2022,1610612740,126.0,0.484,0.926,...,25.0,46.0,1610612759,117.0,0.478,0.815,0.321,23.0,44.0,1
1,2022-12-22,22200478,Final,1610612762,1610612764,2022,1610612762,120.0,0.488,0.952,...,16.0,40.0,1610612764,112.0,0.561,0.765,0.333,20.0,37.0,1
2,2022-12-21,22200466,Final,1610612739,1610612749,2022,1610612739,114.0,0.482,0.786,...,22.0,37.0,1610612749,106.0,0.470,0.682,0.433,20.0,46.0,1
3,2022-12-21,22200467,Final,1610612755,1610612765,2022,1610612755,113.0,0.441,0.909,...,27.0,49.0,1610612765,93.0,0.392,0.735,0.261,15.0,46.0,1
4,2022-12-21,22200468,Final,1610612737,1610612741,2022,1610612737,108.0,0.429,1.000,...,22.0,47.0,1610612741,110.0,0.500,0.773,0.292,20.0,47.0,0


In [4]:
df.columns

Index(['GAME_DATE_EST', 'GAME_ID', 'GAME_STATUS_TEXT', 'HOME_TEAM_ID',
       'VISITOR_TEAM_ID', 'SEASON', 'TEAM_ID_home', 'PTS_home', 'FG_PCT_home',
       'FT_PCT_home', 'FG3_PCT_home', 'AST_home', 'REB_home', 'TEAM_ID_away',
       'PTS_away', 'FG_PCT_away', 'FT_PCT_away', 'FG3_PCT_away', 'AST_away',
       'REB_away', 'HOME_TEAM_WINS'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26651 entries, 0 to 26650
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   GAME_DATE_EST     26651 non-null  object 
 1   GAME_ID           26651 non-null  int64  
 2   GAME_STATUS_TEXT  26651 non-null  object 
 3   HOME_TEAM_ID      26651 non-null  int64  
 4   VISITOR_TEAM_ID   26651 non-null  int64  
 5   SEASON            26651 non-null  int64  
 6   TEAM_ID_home      26651 non-null  int64  
 7   PTS_home          26552 non-null  float64
 8   FG_PCT_home       26552 non-null  float64
 9   FT_PCT_home       26552 non-null  float64
 10  FG3_PCT_home      26552 non-null  float64
 11  AST_home          26552 non-null  float64
 12  REB_home          26552 non-null  float64
 13  TEAM_ID_away      26651 non-null  int64  
 14  PTS_away          26552 non-null  float64
 15  FG_PCT_away       26552 non-null  float64
 16  FT_PCT_away       26552 non-null  float6

## Data Understanding

تحتوي بيانات NBA على 26651 مباراة و 21 عمود.

كل صف يمثل مباراة واحدة بين فريقين: الفريق صاحب الأرض (Home) والفريق الضيف (Away).

تشمل البيانات إحصائيات مهمة لكل فريق مثل:
- النقاط (PTS)
- نسب التصويب (FG%, FT%, 3PT%)
- التمريرات الحاسمة (AST)
- الريباوند (REB)

توجد أعمدة منفصلة لكل من الفريق المضيف (Home) والفريق الضيف (Away)، مما يسمح بمقارنة الأداء بين الفريقين في نفس المباراة.

كما يحتوي العمود (HOME_TEAM_WINS) على نتيجة المباراة، حيث:
- 1 = فوز الفريق صاحب الأرض
- 0 = خسارة الفريق صاحب الأرض

من خلال فحص البيانات باستخدام df.isnull().sum()، نلاحظ وجود 99 قيمة مفقودة (Missing Values) في بعض الأعمدة الخاصة بإحصائيات الأداء.

تمثل هذه القيم المفقودة نسبة صغيرة جدًا من إجمالي البيانات (أقل من 1%)، لذلك سيتم التعامل معها في مرحلة Data Cleaning.

In [6]:
df.describe()

,GAME_ID,HOME_TEAM_ID,VISITOR_TEAM_ID,SEASON,TEAM_ID_home,PTS_home,FG_PCT_home,FT_PCT_home,FG3_PCT_home,AST_home,REB_home,TEAM_ID_away,PTS_away,FG_PCT_away,FT_PCT_away,FG3_PCT_away,AST_away,REB_away,HOME_TEAM_WINS
count,2.665100e+04,2.665100e+04,2.665100e+04,26651.000000,2.665100e+04,26552.000000,26552.000000,26552.000000,26552.000000,26552.000000,26552.000000,2.665100e+04,26552.000000,26552.000000,26552.000000,26552.000000,26552.000000,26552.000000,26651.000000
mean,2.175487e+07,1.610613e+09,1.610613e+09,2012.113879,1.610613e+09,103.455898,0.460735,0.760377,0.356023,22.823441,43.374284,1.610613e+09,100.639876,0.449732,0.758816,0.349489,21.496271,42.113249,0.587032
std,5.570189e+06,8.638670e+00,8.659299e+00,5.587031,8.638670e+00,13.283370,0.056676,0.100677,0.111164,5.193308,6.625769,8.659299e+00,13.435868,0.055551,0.103429,0.109441,5.160596,6.533039,0.492376
min,1.030000e+07,1.610613e+09,1.610613e+09,2003.000000,1.610613e+09,36.000000,0.250000,0.143000,0.000000,6.000000,15.000000,1.610613e+09,33.000000,0.244000,0.143000,0.000000,4.000000,19.000000,0.000000
25%,2.070001e+07,1.610613e+09,1.610613e+09,2007.000000,1.610613e+09,94.000000,0.422000,0.697000,0.286000,19.000000,39.000000,1.610613e+09,91.000000,0.412000,0.692000,0.278000,18.000000,38.000000,0.000000
50%,2.120076e+07,1.610613e+09,1.610613e+09,2012.000000,1.610613e+09,103.000000,0.460000,0.765000,0.357000,23.000000,43.000000,1.610613e+09,100.000000,0.449000,0.765000,0.350000,21.000000,42.000000,1.000000
75%,2.180005e+07,1.610613e+09,1.610613e+09,2017.000000,1.610613e+09,112.000000,0.500000,0.833000,0.429000,26.000000,48.000000,1.610613e+09,110.000000,0.487000,0.833000,0.419000,25.000000,46.000000,1.000000
max,5.210021e+07,1.610613e+09,1.610613e+09,2022.000000,1.610613e+09,168.000000,0.684000,1.000000,1.000000,50.000000,72.000000,1.610613e+09,168.000000,0.687000,1.000000,1.000000,46.000000,81.000000,1.000000


In [7]:
df.isnull().sum()

GAME_DATE_EST        0
GAME_ID              0
GAME_STATUS_TEXT     0
HOME_TEAM_ID         0
VISITOR_TEAM_ID      0
SEASON               0
TEAM_ID_home         0
PTS_home            99
FG_PCT_home         99
FT_PCT_home         99
FG3_PCT_home        99
AST_home            99
REB_home            99
TEAM_ID_away         0
PTS_away            99
FG_PCT_away         99
FT_PCT_away         99
FG3_PCT_away        99
AST_away            99
REB_away            99
HOME_TEAM_WINS       0
dtype: int64

In [8]:
df = df.dropna()

In [9]:
df.isnull().sum()

GAME_DATE_EST       0
GAME_ID             0
GAME_STATUS_TEXT    0
HOME_TEAM_ID        0
VISITOR_TEAM_ID     0
SEASON              0
TEAM_ID_home        0
PTS_home            0
FG_PCT_home         0
FT_PCT_home         0
FG3_PCT_home        0
AST_home            0
REB_home            0
TEAM_ID_away        0
PTS_away            0
FG_PCT_away         0
FT_PCT_away         0
FG3_PCT_away        0
AST_away            0
REB_away            0
HOME_TEAM_WINS      0
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df['GAME_DATE_EST'] = pd.to_datetime(df['GAME_DATE_EST'])

In [12]:
df.dtypes

GAME_DATE_EST       datetime64[ns]
GAME_ID                      int64
GAME_STATUS_TEXT            object
HOME_TEAM_ID                 int64
VISITOR_TEAM_ID              int64
SEASON                       int64
TEAM_ID_home                 int64
PTS_home                   float64
FG_PCT_home                float64
FT_PCT_home                float64
FG3_PCT_home               float64
AST_home                   float64
REB_home                   float64
TEAM_ID_away                 int64
PTS_away                   float64
FG_PCT_away                float64
FT_PCT_away                float64
FG3_PCT_away               float64
AST_away                   float64
REB_away                   float64
HOME_TEAM_WINS               int64
dtype: object

In [13]:
df = df.drop(columns=['GAME_ID'])

In [14]:
df.columns

Index(['GAME_DATE_EST', 'GAME_STATUS_TEXT', 'HOME_TEAM_ID', 'VISITOR_TEAM_ID',
       'SEASON', 'TEAM_ID_home', 'PTS_home', 'FG_PCT_home', 'FT_PCT_home',
       'FG3_PCT_home', 'AST_home', 'REB_home', 'TEAM_ID_away', 'PTS_away',
       'FG_PCT_away', 'FT_PCT_away', 'FG3_PCT_away', 'AST_away', 'REB_away',
       'HOME_TEAM_WINS'],
      dtype='object')

In [15]:
df.shape

(26552, 20)

## Data Cleaning

في هذه المرحلة تم تنظيف البيانات وتجهيزها للتحليل.

- تم فحص القيم المفقودة (Missing Values)، وتبين وجود عدد قليل منها، لذلك تم حذف الصفوف التي تحتوي على قيم مفقودة.
- تم التأكد من عدم وجود بيانات مكررة (Duplicates).
- تم تحويل نوع عمود التاريخ (GAME_DATE_EST) إلى صيغة تاريخية مناسبة للتحليل.
- تم حذف عمود (GAME_ID) لأنه يمثل معرفًا فريدًا لكل مباراة ولا يحتوي على معلومات مفيدة للتحليل.

بعد عملية التنظيف أصبحت البيانات جاهزة لمرحلة التحليل الاستكشافي (EDA).

In [16]:
df['PTS_home'].describe()

count    26552.000000
mean       103.455898
std         13.283370
min         36.000000
25%         94.000000
50%        103.000000
75%        112.000000
max        168.000000
Name: PTS_home, dtype: float64

## تحليل نقاط الفريق صاحب الأرض (PTS_home)

- متوسط عدد النقاط هو حوالي 103 نقطة، وهذا يدل على أن الأداء الطبيعي للفرق يكون حول هذا الرقم.
- أعلى عدد نقاط مسجل هو 168 نقطة، وهو رقم مرتفع لكنه ممكن في بعض المباريات القوية.
- أقل عدد نقاط هو 36 نقطة، وهو رقم منخفض جدًا وقد يشير إلى حالة نادرة أو مشكلة في البيانات.

بشكل عام، توزيع النقاط يبدو منطقيًا مع وجود بعض القيم المتطرفة.

In [17]:
import plotly.express as px

In [18]:
px.histogram(df, x='PTS_home', nbins=50)

من خلال تحليل توزيع عدد النقاط للفريق المضيف (PTS_home)، يتضح أن أغلب القيم تتمركز حول 100 نقطة تقريباً، مما يشير إلى أن هذا هو المعدل الطبيعي لتسجيل النقاط في المباريات.

كما نلاحظ أن التوزيع يبدو متوازنًا بدون وجود قيم شاذة أو غير منطقية، وهذا يدل على أن البيانات نظيفة وتعكس الأداء الحقيقي للفرق بشكل جيد.

In [19]:
df[['PTS_home', 'PTS_away']].describe()

,PTS_home,PTS_away
count,26552.000000,26552.000000
mean,103.455898,100.639876
std,13.283370,13.435868
min,36.000000,33.000000
25%,94.000000,91.000000
50%,103.000000,100.000000
75%,112.000000,110.000000
max,168.000000,168.000000


In [20]:
df['HOME_TEAM_WINS'].value_counts()

HOME_TEAM_WINS
1    15645
0    10907
Name: count, dtype: int64

من خلال تحليل نتائج المباريات، نلاحظ أن الفريق المضيف فاز في 15645 مباراة مقابل 10907 خسارة، مما يشير إلى أن الفريق المضيف يمتلك أفضلية واضحة في الفوز.

هذا يدعم فرضية "أفضلية اللعب على أرض الفريق" (Home Advantage)، حيث يحقق الفريق المضيف نتائج أفضل مقارنة بالفريق الضيف.

In [21]:
fig = px.box(
 df,
 y=['PTS_home', 'PTS_away'],
title='Comparison between Home and Away Points'
 )
fig.update_layout(
    title_x=0.5
 )
fig.show()

يوضح الرسم أن توزيع نقاط الفريق المضيف والضيف متقارب، مع تفوق بسيط للفريق المضيف من حيث المتوسط، مما يشير إلى وجود أفضلية محدودة للعب على أرض الفريق.

In [22]:
df[['PTS_home', 'PTS_away', 'HOME_TEAM_WINS']].corr()

,PTS_home,PTS_away,HOME_TEAM_WINS
PTS_home,1.000000,0.488061,0.395810
PTS_away,0.488061,1.000000,-0.414452
HOME_TEAM_WINS,0.395810,-0.414452,1.000000


هناك علاقة إيجابية بين نقاط الفريق صاحب الأرض واحتمالية الفوز، وعلاقة سلبية بين نقاط الفريق الضيف وفوز الفريق صاحب الأرض. كما أن هناك ارتباط متوسط بين نقاط الفريقين، مما يدل على أن المباريات ذات التسجيل العالي تشمل الفريقين معًا

In [23]:
corr = df[['PTS_home', 'PTS_away', 'HOME_TEAM_WINS']].corr()

fig = px.imshow(corr, text_auto=True, title="Correlation Heatmap")
fig.show()

يظهر تحليل الارتباط أن تسجيل الفريق صاحب الأرض يؤثر إيجابيًا على احتمالية الفوز، بينما تسجيل الفريق الضيف يؤثر سلبًا. كما توجد علاقة متوسطة بين نقاط الفريقين، مما يدل على أن بعض المباريات تكون ذات طابع هجومي للطرفين.

In [24]:
fig = px.scatter(df, x='PTS_home', y='PTS_away',
 title='Home vs Away Points')
fig.show()

"يوضح الرسم وجود علاقة إيجابية بين نقاط الفريق صاحب الأرض والفريق الضيف، مما يشير إلى أن المباريات ذات الطابع الهجومي تؤدي إلى تسجيل نقاط عالية لكلا الفريقين، مع وجود أفضلية بسيطة لصالح الفريق صاحب الأرض."